packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,358 @@
+      perVOSpentVoiceCreditsHash,
+      voteOptionTreeDepth
+    );
+
+    totalVotesSquares += tallyResult ** 2;
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function claim(
+    IPayoutStrategy.Claim calldata params
+  ) public override isInitialized whenNotPaused afterVotingDeadline {
+    uint256 amount = getAllocatedAmount(
+      token.balanceOf(address(this)),
+      params.voiceCreditsPerOption,
+      params.totalSpent,
+      params.tallyResult
+    );
+    totalAmount -= amount;
+
+    IRecipientRegistry.Recipient memory recipient = registry.getRecipient(params.index);
+
+    if (claimed[params.index]) {
+      revert AlreadyClaimed();
+    }
+
+    claimed[params.index] = true;
+
+    bool isValid = verifyTallyResult(
+      params.index,
+      params.tallyResult,
+      params.tallyResultProof,
+      params.tallyResultSalt,
+      params.voteOptionTreeDepth,
+      params.spentVoiceCreditsHash,
+      params.perVOSpentVoiceCreditsHash
+    );
+
+    if (!isValid) {
+      revert InvalidTallyVotesProof();
+    }
+
+    token.safeTransfer(recipient.recipient, amount);
+  }
+
+  /// @notice Get allocated token amounts (without verification).
+  /// @param budget The total budget for the recipients
+  /// @param voiceCreditsPerOptions The voice credit options received for recipient
+  /// @param totalSpent The total amount of voice credits spent
+  /// @param results The tally result for recipient
+  function getAllocatedAmounts(
+    uint256 budget,
+    uint256[] calldata voiceCreditsPerOptions,
+    uint256 totalSpent,
+    uint256[] calldata results
+  ) public view returns (uint256[] memory amounts) {
+    uint256 length = results.length;
+    uint256 alpha = calculateAlpha(budget, totalSpent);
+    amounts = new uint256[](length);
+
+    for (uint256 index = 0; index < length; ) {
+      uint256 tallyResult = results[index];
+      uint256 voiceCreditsPerOption = voiceCreditsPerOptions[index];
+
+      uint256 quadratic = alpha * voiceCreditFactor * tallyResult * tallyResult;
+      uint256 totalSpentCredits = voiceCreditFactor * voiceCreditsPerOption;
+      uint256 linearPrecision = ALPHA_PRECISION * totalSpentCredits;
+      uint256 linearAlpha = alpha * totalSpentCredits;
+
+      amounts[index] = ((quadratic + linearPrecision) - linearAlpha) / ALPHA_PRECISION;
+
+      unchecked {
+        index++;
+      }
+    }
+
+    return amounts;
+  }
+
+  /// @notice Get allocated token amount (without verification).
+  /// @param budget The total budget for the recipients
+  /// @param voiceCreditsPerOption The voice credit options received for recipient
+  /// @param totalSpent The total amount of voice credits spent
+  /// @param tallyResult The tally result for recipient
+  function getAllocatedAmount(
+    uint256 budget,
+    uint256 voiceCreditsPerOption,
+    uint256 totalSpent,
+    uint256 tallyResult
+  ) public view returns (uint256) {
+    uint256 alpha = calculateAlpha(budget, totalSpent);
```

Comment:
I believe the alpha should be constant for the calculations to be correct - if not every time someone claims there is a reduced budget (as this looks like the balance of the token in use). So the first to claim will have a larger budget

Reply:
For alpha we don't have `totalVotesSquares` until coordinator adds tally results and, right, budget will no be the same because it's `token.balanceOf(address(this))`

Reply:
but that would create a race condition to whoever claims first to get an advantage over the next ones

Reply:
then claims should not be allowed until we finalize the round by publishing the batches of results to calculate totalVotesSquared and calculate the alpha - https://github.com/clrfund/monorepo/blob/develop/contracts/contracts/FundingRound.sol#L492

Reply:
Ok, so if the budget is a fixed value in this calculation, it makes sense

Reply:
So to avoid creating additional function to store `totalSpent` and alpha, I'll just store it in base tally contract like the tally results. Here is a PR for it https://github.com/privacy-scaling-explorations/maci/pull/1866